### PR TITLE
Microbatch first last batch serial

### LINKED ISSUE #11094: Run pre/post model hooks once per microbatch model

### LINKED ISSUE #11104: [Micorbatch] Ensure log lines related to microbatch batches have `Batch` prepended to them
### Housekeeping

- [X] I am a maintainer of dbt-core

### Short description

Right now logs about batches look something like the following
Screenshot 2024-12-06 at 16 53 26 (2)

Notice how lines about batches `X of 4` have the same format as lines about _nodes_ `X of 3`. This is confusing because there easy way to visually parse what is going on. 

### Acceptance criteria

Make batch log lines distinct to be more easily visually parsable

### Suggested Tests

N/A

### Impact to Other Teams

N/A

### Will backports be required?

Backport to 1.9.latest

### Context

_No response_

### PR DESCRIPTION
**Pull request by MichelleArk**
Resolves #11094
Resolves #11104

### Problem

Model pre & post hooks were running on each batch for a microbatch model

### Solution

* First and last batch are always run in serial
* First batch runs pre-hook
* Last batch runs post-hook
* Account for when first batch = last batch
* If first batch fails, skip remaining batches

### Examples

**When all batches succeed**
Screenshot 2024-12-06 at 19 29 17 (2)

**When the first batch fails**
Screenshot 2024-12-06 at 19 29 33 (2)

**When a middle batch fails**
Screenshot 2024-12-06 at 19 30 49 (2)

### Checklist

- [X] I have read [the contributing guide](https://github.com/dbt-labs/dbt-core/blob/main/CONTRIBUTING.md) and understand what's expected of me.
- [X] I have run this code in development, and it appears to resolve the stated issue.
- [X] This PR includes tests, or tests are not required or relevant for this PR.
- [X] This PR has no interface changes (e.g., macros, CLI, logs, JSON artifacts, config files, adapter interface, etc.) or this PR has already received feedback and approval from Product or DX.
- [X] This PR includes [type annotations](https://docs.python.org/3/library/typing.html) for new and modified functions.

### CODE REVIEW

File: core/dbt/task/run.py [hunk 1]
```diff
@@ -717,30 +715,20 @@ def handle_microbatch_model(
 
         batch_results: List[RunResult] = []
         batch_idx = 0
-
-        # Run first batch runs in serial
-        relation_exists = self._submit_batch(
-            node, relation_exists, batches, batch_idx, batch_results, pool, parallel=False
-        )
-        batch_idx += 1
```
Comment (QMalcolm) [inline core/dbt/task/run.py]: I think it's actually preferable to the first batch separated out. It makes it mirror how the last batch is separated out. We could also have an optional arg `force_sequential` (defaulted to `False`) which would skip the `should_run_in_parallel` check in `_submit_batch`

File: core/dbt/task/run.py [hunk 2]
```diff
@@ -602,15 +602,15 @@ def _has_relation(self, model) -> bool:
...
+    def should_run_in_parallel(self) -> bool:
         if not self.adapter.supports(Capability.MicrobatchConcurrency):
             run_in_parallel = False
-        elif not relation_exists:
+        elif not self.relation_exists:
             # If the relation doesn't exist, we can't run in parallel
             run_in_parallel = False
+        elif self.batch_idx == 0 or self.batch_idx == len(self.batches) - 1:
+            # First and last batch don't run in parallel
+            run_in_parallel = False
```
Comment (QMalcolm) [inline core/dbt/task/run.py:613]: This check could also be skipped if we're instead handling `force_sequential` to determine if we should even check `should_run_in_parallel` in `_submit_batch`. It'd be nice for this function to be less dependent on "where" it is, and I think this check breaks that.

**Review by QMalcolm — COMMENTED** [review]
Woooo! Thank you for doing this work ❤️ I think this will make it so hooks work how people expect, which is incredibly important. I do think there are some changes we can make though to improve the mental model of what happens where for maintainability though.

File: tests/functional/microbatch/test_microbatch.py [hunk 3]
```diff
@@ -905,3 +967,100 @@ def test_microbatch(
...
+            # batch id that should be firing pre-hook
+            if "20200101" in event.data.msg:  # type: ignore
+                assert "pre-hook" in event.data.msg  # type: ignore
+
+            # batch id that should be firing the post-hook
+            if "20200104" in event.data.msg:  # type: ignore
+                assert "post-hook" in event.data.msg  # type: ignore
+
+
+class TestWhenOnlyOneBatchRunBothPostAndPreHooks(BaseMicrobatchTest):
```
Comment (MichelleArk) [inline tests/functional/microbatch/test_microbatch.py:1036]: thank you for adding this one 🙏

**Review by MichelleArk — COMMENTED** [review]
✅ Can't approve my own PR, but thank you so much for taking this over the finish line 🏁 

The rest of the functional changes + tests look great 🚀

**Review by QMalcolm — APPROVED** [review]
Self approval 🙈

**Comment by QMalcolm** [comment]
Tests look good, proceeding with merging

**Pull request merged**